File: bundler/bundler.py
import argparse
import struct
import shutil
import os
import sys

# Magic bytes "GGUF" in hex: 0x47475546
BUNDLE_MAGIC = 0x47475546

def bundle(server_path, model_path, output_path):
    print(f"Bundling {server_path} + {model_path} -> {output_path}")
    
    # Read server
    try:
        with open(server_path, 'rb') as f:
            server_data = f.read()
    except FileNotFoundError:
        print(f"Error: Server file '{server_path}' not found.")
        sys.exit(1)

    # Read model
    try:
        with open(model_path, 'rb') as f:
            model_data = f.read()
    except FileNotFoundError:
        print(f"Error: Model file '{model_path}' not found.")
        sys.exit(1)

    # Calculate offsets
    server_size = len(server_data)
    model_size = len(model_data)
    gguf_offset = server_size
    
    print(f"Server size: {server_size} bytes")
    print(f"Model size: {model_size} bytes")
    print(f"GGUF offset: {gguf_offset}")

    # Write output
    with open(output_path, 'wb') as f:
        f.write(server_data)
        f.write(model_data)
        
        # Write footer
        # struct BundleFooter { uint64_t offset; uint64_t size; uint32_t magic; };
        # Use little-endian standard
        footer = struct.pack('<QQI', gguf_offset, model_size, BUNDLE_MAGIC)
        f.write(footer)

    # Make executable
    st = os.stat(output_path)
    os.chmod(output_path, st.st_mode | 0o111)
    
    print(f"Successfully created bundled executable: {output_path}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Bundle a server executable and a GGUF model into a single binary.")
    parser.add_argument("--server", required=True, help="Path to the server executable")
    parser.add_argument("--model", required=True, help="Path to the GGUF model file")
    parser.add_argument("--out", required=True, help="Path for the output bundled executable")
    
    args = parser.parse_args()
    bundle(args.server, args.model, args.out)

File: bundler/bundler.cpp
#include <cstdint>
#include <cstring>
#include <fstream>
#include <iostream>
#include <string>
#include <sys/stat.h>
#include <vector>

#ifdef _WIN32
#include <windows.h>
#define PATH_SEP '\\'
#else
#include <sys/stat.h>
#include <unistd.h>
#define PATH_SEP '/'
#endif

#pragma pack(push, 1)
struct BundleFooter {
  uint64_t model_offset;
  uint64_t model_size;
  uint32_t magic; // 0x47475546 (GGUF)
};
#pragma pack(pop)

bool set_executable(const char *path) {
#ifdef _WIN32
  return true;
#else
  struct stat st;
  if (stat(path, &st) != 0)
    return false;
  return chmod(path, st.st_mode | S_IXUSR | S_IXGRP | S_IXOTH) == 0;
#endif
}

int main(int argc, char *argv[]) {
  if (argc < 3) {
    std::cerr << "Usage: " << argv[0]
              << " <server_exe> <model_gguf> [output_file.com]" << std::endl;
    return 1;
  }

  std::string server_path = argv[1];
  std::string model_path = argv[2];
  std::string output_path = (argc > 3) ? argv[3] : "";

  if (output_path.empty()) {
    output_path = model_path;
    size_t last_dot = output_path.find_last_of('.');
    if (last_dot != std::string::npos)
      output_path = output_path.substr(0, last_dot);
    output_path += ".baremetallama";
  } else if (output_path.find(".baremetallama") == std::string::npos &&
             output_path.find(".") == std::string::npos) {
    output_path += ".baremetallama";
  }

  std::ifstream server_infile(server_path, std::ios::binary);
  std::ifstream model_infile(model_path, std::ios::binary);

  if (!server_infile || !model_infile) {
    std::cerr << "Error opening input files." << std::endl;
    return 1;
  }

  // Get server binary size
  server_infile.seekg(0, std::ios::end);
  uint64_t server_size = server_infile.tellg();
  server_infile.seekg(0, std::ios::beg);

  std::ofstream output_file(output_path, std::ios::binary);
  if (!output_file) {
    std::cerr << "Error opening output file: " << output_path << std::endl;
    return 1;
  }

  // Write server binary (Payload at START of file for PE compatibility)
  output_file << server_infile.rdbuf();

  // Align model to 64KB boundary for cross-platform mmap compatibility (Windows
  // requirements)
  uint64_t cur = output_file.tellp();
  uint64_t model_offset = (cur + 65535) & ~65535;
  if (model_offset > cur) {
    std::vector<char> padding(model_offset - cur, 0);
    output_file.write(padding.data(), padding.size());
  }

  // Write model data
  output_file << model_infile.rdbuf();
  uint64_t model_size = (uint64_t)output_file.tellp() - model_offset;

  // Append footer
  BundleFooter footer;
  footer.model_offset = model_offset;
  footer.model_size = model_size;
  footer.magic = 0x47475546; // GGUF magic in footer too

  output_file.write(reinterpret_cast<const char *>(&footer), sizeof(footer));

  output_file.close();
  set_executable(output_path.c_str());

  std::cout << "Successfully created Universal BareMetalLlama: " << output_path
            << std::endl;
  std::cout << "Model data offset: " << model_offset << ", size: " << model_size
            << std::endl;
  std::cout << "This file is now a valid APE binary and should run on Linux, "
               "Windows, Mac."
            << std::endl;

  return 0;
}
